File: dotnet/src/Pactman.NonprofitCheckPlus/Http/Transport.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Configuration;
using Pactman.NonprofitCheckPlus.Exceptions;
using Pactman.NonprofitCheckPlus.Internal;
using Pactman.NonprofitCheckPlus.Models;

namespace Pactman.NonprofitCheckPlus.Http
{
    /// <summary>
    /// HTTP transport: authentication headers, per-attempt deadlines, retries with
    /// jittered backoff, <c>Retry-After</c> handling, and mapping responses to the error
    /// taxonomy.
    /// </summary>
    /// <remarks>
    /// The API key is held in a delegate and written into the <c>Authorization</c> header
    /// at send time. It is never stored as a field, so no diagnostic that walks this
    /// object can reach it.
    /// <para>Internal. Nothing here is part of the public API.</para>
    /// </remarks>
    internal sealed class Transport
    {
        private readonly Func<string> _credential;
        private readonly ClientConfig _config;
        private readonly HttpClient _httpClient;
        private readonly Func<TimeSpan, CancellationToken, Task> _delay;
        private readonly Func<double> _random;
        private readonly Func<double> _monotonic;
        private readonly object _throttleLock = new object();

        private double _nextRequestAt;

        /// <summary>Initializes the transport.</summary>
        /// <param name="credential">Returns the <c>Authorization</c> header value.</param>
        /// <param name="config">The resolved client configuration.</param>
        /// <param name="httpClient">Where to send through.</param>
        /// <param name="hooks">Test seam for the clock. Never set in production.</param>
        internal Transport(
            Func<string> credential,
            ClientConfig config,
            HttpClient httpClient,
            TransportHooks? hooks = null)
        {
            _credential = credential;
            _config = config;
            _httpClient = httpClient;
            _delay = hooks?.Delay ?? ((duration, token) => Task.Delay(duration, token));
            _random = hooks?.Random ?? DefaultRandom;
            _monotonic = hooks?.Monotonic ?? DefaultMonotonic;
        }

        /// <summary>Sends a request, retrying it under the resolved policy.</summary>
        /// <param name="method"><c>GET</c> or <c>POST</c>.</param>
        /// <param name="path">Path to append to the configured base URL.</param>
        /// <param name="body">JSON-encoded when present.</param>
        /// <param name="timeout">Deadline for one attempt; the client's when omitted.</param>
        /// <param name="retry">Policy for this request; the client's when omitted.</param>
        /// <param name="headers">Extra headers. Cannot override <c>Authorization</c>.</param>
        /// <param name="cancellationToken">Bounds the call as a whole, retries included.</param>
        /// <returns>The parsed response.</returns>
        /// <exception cref="PactmanApiException">A non-2xx response the policy will not retry.</exception>
        /// <exception cref="PactmanTimeoutException">The deadline expired on the final attempt.</exception>
        /// <exception cref="PactmanNetworkException">No response was produced on the final attempt.</exception>
        /// <exception cref="OperationCanceledException"><paramref name="cancellationToken"/> was cancelled.</exception>
        internal async Task<TransportResponse> SendAsync(
            HttpMethod method,
            string path,
            IReadOnlyList<string>? body,
            TimeSpan? timeout,
            RetryOptions? retry,
            IReadOnlyDictionary<string, string>? headers,
            CancellationToken cancellationToken)
        {
            var policy = retry ?? _config.Retry;
            var attemptTimeout = timeout ?? _config.Timeout;
            var url = _config.BaseUrl + path;
            var payload = body is null ? null : JsonSerializer.Serialize(body);
            var attempts = 0;

            while (true)
            {
                attempts++;
                cancellationToken.ThrowIfCancellationRequested();
                await ThrottleAsync(cancellationToken).ConfigureAwait(false);

                int status;
                string? requestId;
                ResponseBody parsed;
                TimeSpan? retryAfter;

                try
                {
                    using var attemptCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                    attemptCancellation.CancelAfter(attemptTimeout);

                    using var request = BuildRequest(method, url, payload, headers);
                    using var response = await _httpClient
                        .SendAsync(request, HttpCompletionOption.ResponseContentRead, attemptCancellation.Token)
                        .ConfigureAwait(false);

                    status = (int)response.StatusCode;
                    requestId = ReadRequestId(response);
                    retryAfter = ReadRetryAfter(response, DateTimeOffset.UtcNow);
                    parsed = ParseBody(await ReadBodyAsync(response, attemptCancellation.Token).ConfigureAwait(false));
                }
                catch (Exception failure) when (IsTransportFailure(failure, cancellationToken))
                {
                    // The caller's own cancellation surfaces as OperationCanceledException
                    // from the filter above declining it, so "I gave up" never arrives
                    // dressed as "the API was too slow".
                    var error = Classify(failure, attemptTimeout, attempts);

                    if (attempts > policy.MaxRetries)
                    {
                        throw error;
                    }

                    await _delay(ComputeRetryDelay(attempts, policy, null, _random), cancellationToken)
                        .ConfigureAwait(false);

                    continue;
                }

                if (status >= 200 && status < 300)
                {
                    return new TransportResponse(status, requestId, parsed, attempts);
                }

                var apiError = PactmanApiException.FromStatus(BuildApiErrorInit(
                    status: status,
                    parsed: parsed,
                    requestId: requestId,
                    retryAfter: retryAfter,
                    attempts: attempts));

                if (attempts > policy.MaxRetries || !policy.IsRetryableStatus(status))
                {
                    throw apiError;
                }

                await _delay(ComputeRetryDelay(attempts, policy, retryAfter, _random), cancellationToken)
                    .ConfigureAwait(false);
            }
        }

        /// <summary>
        /// Delay before the next attempt.
        /// </summary>
        /// <remarks>
        /// A valid <c>Retry-After</c> wins outright. Otherwise the delay grows
        /// exponentially from <see cref="RetryOptions.InitialDelay"/>, is capped at
        /// <see cref="RetryOptions.MaxDelay"/>, and — with jitter on — is randomized across
        /// the whole range so concurrent clients spread out.
        /// </remarks>
        /// <param name="attempt">Which attempt just failed, counting from one.</param>
        /// <param name="retry">The policy in force.</param>
        /// <param name="retryAfter">The server's <c>Retry-After</c>, when it sent one.</param>
        /// <param name="random">Returns a value in <c>[0, 1)</c>.</param>
        /// <returns>How long to wait before the next attempt.</returns>
        internal static TimeSpan ComputeRetryDelay(
            int attempt,
            RetryOptions retry,
            TimeSpan? retryAfter,
            Func<double>? random = null)
        {
            if (retry.RespectRetryAfter && retryAfter.HasValue && retryAfter.Value >= TimeSpan.Zero)
            {
                return Round(retryAfter.Value);
            }

            var exponential = retry.InitialDelay.TotalSeconds * Math.Pow(retry.BackoffFactor, attempt - 1);
            var capped = Math.Min(exponential, retry.MaxDelay.TotalSeconds);

            if (!retry.Jitter)
            {
                return Round(TimeSpan.FromSeconds(capped));
            }

            random ??= DefaultRandom;

            return Round(TimeSpan.FromSeconds(random() * capped));
        }

        /// <summary>Reads <c>Retry-After</c> as either a delay or an HTTP date.</summary>
        /// <param name="response">The response to read.</param>
        /// <param name="now">The current time, against which a date is measured.</param>
        /// <returns>The delay the server asked for, or <see langword="null"/>.</returns>
        internal static TimeSpan? ReadRetryAfter(HttpResponseMessage response, DateTimeOffset now)
        {
            var retryAfter = response.Headers.RetryAfter;

            if (retryAfter is null)
            {
                return null;
            }

            if (retryAfter.Delta.HasValue)
            {
                return retryAfter.Delta.Value < TimeSpan.Zero ? (TimeSpan?)null : retryAfter.Delta.Value;
            }

            if (retryAfter.Date.HasValue)
            {
                var delta = retryAfter.Date.Value - now;

                return delta < TimeSpan.Zero ? TimeSpan.Zero : delta;
            }

            return null;
        }

        /// <summary>Reads the correlation identifier from the response headers.</summary>
        /// <param name="response">The response to read.</param>
        /// <returns>The first correlation header the server sent, or <see langword="null"/>.</returns>
        internal static string? ReadRequestId(HttpResponseMessage response)
        {
            foreach (var name in new[] { "x-request-id", "x-correlation-id", "request-id" })
            {
                if (response.Headers.TryGetValues(name, out var values))
                {
                    foreach (var value in values)
                    {
                        return value;
                    }
                }
            }

            return null;
        }

        /// <summary>
        /// Parses a response body, preserving unparseable text as evidence.
        /// </summary>
        /// <param name="text">The raw body.</param>
        /// <returns>The parsed JSON, the raw text, or an empty body.</returns>
        internal static ResponseBody ParseBody(string? text)
        {
            if (string.IsNullOrWhiteSpace(text))
            {
                return ResponseBody.Empty;
            }

            JsonElement parsed;

            try
            {
                parsed = JsonValues.Parse(text!);
            }
            catch (JsonException)
            {
                // Preserve whatever the server sent; an unparseable body is still evidence.
                return ResponseBody.FromText(text!);
            }

            // A bare JSON scalar is not an envelope; keep the text as evidence.
            return parsed.ValueKind == JsonValueKind.Object || parsed.ValueKind == JsonValueKind.Array
                ? ResponseBody.FromJson(parsed)
                : ResponseBody.FromText(text!);
        }

        /// <summary>Normalizes the envelope's <c>errors</c> into a list.</summary>
        /// <param name="errors">The envelope's <c>errors</c> value, whatever shape it took.</param>
        /// <returns>The item-level failures, or an empty list.</returns>
        internal static IReadOnlyList<ApiErrorDetail> NormalizeApiErrors(JsonElement? errors)
        {
            var details = new List<ApiErrorDetail>();

            if (!errors.HasValue)
            {
                return details;
            }

            var element = errors.Value;

            switch (element.ValueKind)
            {
                case JsonValueKind.String:
                    var reason = element.GetString();

                    if (!string.IsNullOrWhiteSpace(reason))
                    {
                        details.Add(ApiErrorDetail.FromReason(reason!));
                    }

                    break;

                case JsonValueKind.Array:
                    foreach (var entry in element.EnumerateArray())
                    {
                        if (entry.ValueKind == JsonValueKind.Object)
                        {
                            details.Add(ApiErrorDetail.FromJson(entry));
                        }
                    }

                    break;

                case JsonValueKind.Object:
                    // A single error object, rather than a list of them.
                    details.Add(ApiErrorDetail.FromJson(element));
                    break;

                default:
                    break;
            }

            return details;
        }

        /// <summary>Collects the response metadata an API error carries.</summary>
        /// <param name="status">HTTP status of the response.</param>
        /// <param name="parsed">The parsed body.</param>
        /// <param name="requestId">Correlation identifier, when present.</param>
        /// <param name="retryAfter">The server's <c>Retry-After</c>, when it sent one.</param>
        /// <param name="attempts">How many attempts were made.</param>
        /// <returns>The metadata bundle for an API exception.</returns>
        internal static ApiErrorInit BuildApiErrorInit(
            int status,
            ResponseBody parsed,
            string? requestId,
            TimeSpan? retryAfter,
            int attempts)
        {
            var envelope = parsed.Envelope;
            var hasEnvelope = envelope.ValueKind == JsonValueKind.Object;

            var apiErrors = NormalizeApiErrors(
                hasEnvelope && envelope.TryGetProperty("errors", out var errors) ? errors : (JsonElement?)null);

            var reasons = new List<string>();

            foreach (var detail in apiErrors)
            {
                if (!string.IsNullOrWhiteSpace(detail.Reason))
                {
                    reasons.Add(detail.Reason!);
                }
            }

            string? apiMessage;

            if (reasons.Count > 0)
            {
                apiMessage = string.Join("; ", reasons);
            }
            else if (hasEnvelope
                && envelope.TryGetProperty("message", out var message)
                && message.ValueKind == JsonValueKind.String)
            {
                apiMessage = message.GetString();
            }
            else if (!string.IsNullOrWhiteSpace(parsed.Text))
            {
                var trimmed = parsed.Text!.Trim();
                apiMessage = trimmed.Length > 500 ? trimmed.Substring(0, 500) : trimmed;
            }
            else
            {
                apiMessage = null;
            }

            int? apiCode = null;

            if (hasEnvelope && envelope.TryGetProperty("code", out var code))
            {
                apiCode = JsonValues.AsInt32(code);
            }

            return new ApiErrorInit(
                status: status,
                apiMessage: apiMessage,
                apiCode: apiCode,
                apiErrors: apiErrors,
                requestId: requestId,
                retryAfter: retryAfter,
                raw: parsed,
                attempts: attempts);
        }

        private static bool IsTransportFailure(Exception failure, CancellationToken cancellationToken)
        {
            // A cancellation the caller asked for is theirs to see, untranslated.
            if (cancellationToken.IsCancellationRequested)
            {
                return false;
            }

            return failure is HttpRequestException
                || failure is OperationCanceledException
                || failure is System.IO.IOException;
        }

        private static PactmanException Classify(Exception failure, TimeSpan timeout, int attempts)
        {
            if (failure is OperationCanceledException)
            {
                return new PactmanTimeoutException(
                    string.Format(
                        CultureInfo.InvariantCulture,
                        "The request timed out after {0}s.",
                        timeout.TotalSeconds),
                    timeout,
                    attempts,
                    failure);
            }

            return new PactmanNetworkException(
                "The request to the Pactman API failed: " + failure.Message,
                attempts,
                failure);
        }

        private static TimeSpan Round(TimeSpan value) =>
            TimeSpan.FromMilliseconds(Math.Round(value.TotalMilliseconds));

        private static double DefaultMonotonic() => Stopwatch.GetTimestamp() / (double)Stopwatch.Frequency;

#if NET8_0_OR_GREATER
        private static double DefaultRandom() => System.Random.Shared.NextDouble();
#else
        [ThreadStatic]
        private static Random? _threadRandom;

        private static double DefaultRandom() =>
            (_threadRandom ??= new Random(Guid.NewGuid().GetHashCode())).NextDouble();
#endif

        private static async Task<string> ReadBodyAsync(HttpResponseMessage response, CancellationToken cancellationToken)
        {
#if NET8_0_OR_GREATER
            return await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
#else
            cancellationToken.ThrowIfCancellationRequested();

            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
#endif
        }

        private HttpRequestMessage BuildRequest(
            HttpMethod method,
            string url,
            string? payload,
            IReadOnlyDictionary<string, string>? perRequest)
        {
            var request = new HttpRequestMessage(method, url);

            if (payload != null)
            {
                request.Content = new StringContent(payload, Encoding.UTF8, "application/json");
            }

            var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

            foreach (var header in _config.DefaultHeaders)
            {
                headers[header.Key] = header.Value;
            }

            if (perRequest != null)
            {
                foreach (var header in perRequest)
                {
                    headers[header.Key] = header.Value;
                }
            }

            // Set last so neither the client defaults nor a per-request header can
            // displace the credential or misdeclare the payload.
            headers["Accept"] = "application/json";
            headers["User-Agent"] = _config.UserAgent;
            headers["Authorization"] = _credential();

            foreach (var header in headers)
            {
                if (request.Headers.TryAddWithoutValidation(header.Key, header.Value))
                {
                    continue;
                }

                // Content-Type and friends belong on the content, not the request.
                request.Content?.Headers.Remove(header.Key);
                request.Content?.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            return request;
        }

        /// <summary>Spaces requests when <see cref="ClientConfig.MaxRequestsPerSecond"/> is configured.</summary>
        private Task ThrottleAsync(CancellationToken cancellationToken)
        {
            var limit = _config.MaxRequestsPerSecond;

            if (limit is null)
            {
                return Task.CompletedTask;
            }

            double wait;

            // A client is meant to be shared across threads, so the schedule is claimed
            // under a lock. The waiting itself happens outside it.
            lock (_throttleLock)
            {
                var interval = 1.0 / limit.Value;
                var now = _monotonic();
                var scheduledAt = Math.Max(now, _nextRequestAt);
                _nextRequestAt = scheduledAt + interval;
                wait = scheduledAt - now;
            }

            return wait > 0 ? _delay(TimeSpan.FromSeconds(wait), cancellationToken) : Task.CompletedTask;
        }
    }
}

File: go/pactman/transport.go
package pactman

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"math"
	"math/rand/v2"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

// hooks are the clock and random source, substituted by the tests.
type hooks struct {
	// sleep waits for d, or returns ctx's error if it ends first.
	sleep func(ctx context.Context, d time.Duration) error
	// random returns a value in [0, 1), for backoff jitter.
	random func() float64
	now    func() time.Time
}

func defaultHooks() hooks {
	return hooks{sleep: sleepContext, random: rand.Float64, now: time.Now}
}

// transport issues authenticated requests: headers, per-attempt timeouts,
// cancellation, retries with jittered backoff, Retry-After, and mapping
// responses onto the error taxonomy.
//
// The API key is written into the Authorization header at send time and is
// never stored on a request record, an error, or any diagnostic output.
type transport struct {
	apiKey    secret
	config    clientConfig
	hooks     hooks
	userAgent string

	// mu guards nextRequestAt: a client is shared across goroutines.
	mu            sync.Mutex
	nextRequestAt time.Time
}

// response is a successful HTTP response and the metadata a result needs.
type response struct {
	status    int
	requestID string
	body      []byte
	attempts  int
}

// requestSettings resolves a call's options over the client's settings.
func (t *transport) requestSettings(opts []RequestOption) (requestConfig, error) {
	settings := requestConfig{
		timeout: t.config.timeout,
		retry:   t.config.retry.clone(),
		headers: http.Header{},
	}

	for _, opt := range opts {
		if opt == nil {
			continue
		}

		if err := opt.applyRequest(&settings); err != nil {
			return requestConfig{}, err
		}
	}

	return settings, nil
}

func (t *transport) send(ctx context.Context, method, path string, payload []byte, settings requestConfig) (*response, error) {
	url := t.config.baseURL + path
	retry := settings.retry

	for attempt := 1; ; attempt++ {
		if err := ctx.Err(); err != nil {
			return nil, aborted(attempt-1, err)
		}

		if err := t.throttle(ctx); err != nil {
			return nil, aborted(attempt-1, err)
		}

		res, err := t.attempt(ctx, method, url, payload, settings, attempt)

		if err != nil {
			// A timeout or a transport failure is retried; the caller giving up is not.
			if attempt > retry.MaxRetries || ctx.Err() != nil {
				return nil, err
			}

			if err := t.waitBeforeRetry(ctx, attempt, retry, nil); err != nil {
				return nil, aborted(attempt, err)
			}

			continue
		}

		if res.StatusCode >= 200 && res.StatusCode < 300 {
			return &response{
				status:    res.StatusCode,
				requestID: readRequestID(res.Header),
				body:      res.body,
				attempts:  attempt,
			}, nil
		}

		retryAfter := readRetryAfter(res.Header.Get("Retry-After"), t.hooks.now())
		apiErr := buildAPIError(res.StatusCode, res.body, readRequestID(res.Header), retryAfter, attempt)

		if attempt > retry.MaxRetries || !retry.retries(res.StatusCode) {
			return nil, apiErr
		}

		if err := t.waitBeforeRetry(ctx, attempt, retry, retryAfter); err != nil {
			return nil, aborted(attempt, err)
		}
	}
}

// httpResult is a response whose body has been read in full.
type httpResult struct {
	*http.Response
	body []byte
}

// attempt makes one request. The per-attempt timeout covers reading the body
// as well as waiting for the headers.
func (t *transport) attempt(ctx context.Context, method, url string, payload []byte, settings requestConfig, attempt int) (*httpResult, error) {
	attemptCtx, cancel := context.WithTimeout(ctx, settings.timeout)
	defer cancel()

	var body io.Reader
	if payload != nil {
		body = bytes.NewReader(payload)
	}

	req, err := http.NewRequestWithContext(attemptCtx, method, url, body)
	if err != nil {
		return nil, &NetworkError{Attempts: attempt, msg: "the request could not be built", cause: err}
	}

	req.Header = t.headers(settings.headers, payload != nil)

	res, err := t.config.httpClient.Do(req)
	if err == nil {
		data, readErr := io.ReadAll(res.Body)
		res.Body.Close()

		if readErr == nil {
			return &httpResult{Response: res, body: data}, nil
		}

		err = readErr
	}

	// The caller's context is checked first: when its deadline is the earlier
	// one, both contexts report it, and it is the caller who gave up.
	if ctx.Err() != nil {
		return nil, aborted(attempt, ctx.Err())
	}

	if errors.Is(attemptCtx.Err(), context.DeadlineExceeded) {
		return nil, &TimeoutError{Timeout: settings.timeout, Attempts: attempt}
	}

	return nil, &NetworkError{Attempts: attempt, msg: "the request to the Pactman API failed", cause: err}
}

// Header names this SDK owns. A caller's value for one is dropped rather than
// sent beside the real one.
var reservedHeaders = []string{"Authorization", "Accept", "User-Agent", "Content-Type"}

func (t *transport) headers(perRequest http.Header, hasBody bool) http.Header {
	headers := t.config.headers.Clone()

	for name, values := range perRequest {
		headers[name] = append([]string(nil), values...)
	}

	for _, name := range reservedHeaders {
		headers.Del(name)
	}

	headers.Set("Accept", "application/json")
	headers.Set("User-Agent", t.userAgent)
	headers.Set("Authorization", "Bearer "+t.apiKey.reveal())

	if hasBody {
		headers.Set("Content-Type", "application/json")
	}

	return headers
}

// throttle spaces requests when a request-per-second ceiling is configured.
func (t *transport) throttle(ctx context.Context) error {
	limit := t.config.maxRequestsPerSecond
	if limit == 0 {
		return nil
	}

	interval := time.Duration(float64(time.Second) / limit)

	t.mu.Lock()
	now := t.hooks.now()
	scheduled := now

	if t.nextRequestAt.After(now) {
		scheduled = t.nextRequestAt
	}

	t.nextRequestAt = scheduled.Add(interval)
	t.mu.Unlock()

	if wait := scheduled.Sub(now); wait > 0 {
		return t.hooks.sleep(ctx, wait)
	}

	return nil
}

func (t *transport) waitBeforeRetry(ctx context.Context, attempt int, retry RetryPolicy, retryAfter *time.Duration) error {
	return t.hooks.sleep(ctx, computeRetryDelay(attempt, retry, retryAfter, t.hooks.random))
}

// computeRetryDelay is the delay before the next attempt.
//
// A usable Retry-After wins outright. Otherwise the delay grows exponentially
// from InitialDelay, is capped at MaxDelay, and — with jitter on — is
// randomized across the whole range so concurrent clients spread out. Delays
// are whole milliseconds.
func computeRetryDelay(attempt int, retry RetryPolicy, retryAfter *time.Duration, random func() float64) time.Duration {
	if retry.RespectRetryAfter && retryAfter != nil && *retryAfter >= 0 {
		return roundToMillisecond(float64(*retryAfter))
	}

	exponential := float64(retry.InitialDelay) * math.Pow(retry.BackoffFactor, float64(attempt-1))
	capped := math.Min(exponential, float64(retry.MaxDelay))

	if retry.Jitter {
		capped *= random()
	}

	return roundToMillisecond(capped)
}

func roundToMillisecond(nanoseconds float64) time.Duration {
	return time.Duration(math.Round(nanoseconds/float64(time.Millisecond))) * time.Millisecond
}

// readRetryAfter reads Retry-After as a delay in seconds or an HTTP date, or
// returns nil when there is no usable value.
func readRetryAfter(header string, now time.Time) *time.Duration {
	raw := strings.TrimSpace(header)
	if raw == "" {
		return nil
	}

	if seconds, err := strconv.ParseFloat(raw, 64); err == nil {
		if math.IsNaN(seconds) || math.IsInf(seconds, 0) || seconds < 0 {
			return nil
		}

		delay := time.Duration(math.MaxInt64)
		if seconds < float64(math.MaxInt64)/float64(time.Second) {
			delay = time.Duration(seconds * float64(time.Second))
		}

		return &delay
	}

	when, ok := parseHTTPDate(raw)
	if !ok {
		return nil
	}

	delay := max(when.Sub(now), 0)

	return &delay
}

// parseHTTPDate reads the IMF-fixdate form HTTP specifies, its two obsolete
// forms, and RFC 1123 with a zone other than GMT, which servers send too.
func parseHTTPDate(raw string) (time.Time, bool) {
	if when, err := http.ParseTime(raw); err == nil {
		return when, true
	}

	for _, layout := range []string{time.RFC1123, time.RFC1123Z} {
		if when, err := time.Parse(layout, raw); err == nil {
			return when, true
		}
	}

	return time.Time{}, false
}

func readRequestID(headers http.Header) string {
	for _, name := range []string{"X-Request-Id", "X-Correlation-Id", "Request-Id"} {
		if value := headers.Get(name); value != "" {
			return value
		}
	}

	return ""
}

// buildAPIError maps an error response onto an *APIError, keeping its
// metadata even when the body is not the JSON envelope.
func buildAPIError(status int, body []byte, requestID string, retryAfter *time.Duration, attempts int) *APIError {
	apiErr := &APIError{
		Status:     status,
		RequestID:  requestID,
		RetryAfter: retryAfter,
		Body:       body,
		Attempts:   attempts,
		APIErrors:  []APIErrorDetail{},
	}

	trimmed := bytes.TrimSpace(body)

	if fields, err := parseObject(trimmed); err == nil {
		var envelope Envelope

		envelope.Fields = fields
		decodeDeclared(&envelope, fields)

		if envelope.Errors != nil {
			apiErr.APIErrors = envelope.Errors
		}

		apiErr.APICode = envelope.Code

		var reasons []string

		for _, detail := range apiErr.APIErrors {
			if strings.TrimSpace(detail.Reason) != "" {
				reasons = append(reasons, detail.Reason)
			}
		}

		switch {
		case len(reasons) > 0:
			apiErr.APIMessage = strings.Join(reasons, "; ")
		case envelope.Message != nil:
			apiErr.APIMessage = *envelope.Message
		}
	} else if text, ok := bodyText(trimmed); ok {
		apiErr.APIMessage = truncateRunes(text, 500)
	}

	return errorFromStatus(apiErr)
}

// bodyText is a body that is text rather than a JSON object: a JSON string's
// value, or the body itself when it is not JSON at all.
func bodyText(trimmed []byte) (string, bool) {
	if len(trimmed) == 0 {
		return "", false
	}

	if !json.Valid(trimmed) {
		return string(trimmed), true
	}

	var text string
	if json.Unmarshal(trimmed, &text) == nil && strings.TrimSpace(text) != "" {
		return strings.TrimSpace(text), true
	}

	return "", false
}

func truncateRunes(text string, limit int) string {
	runes := []rune(text)
	if len(runes) <= limit {
		return text
	}

	return string(runes[:limit])
}

func aborted(attempts int, cause error) *NetworkError {
	return &NetworkError{Attempts: attempts, msg: "the request was aborted by the caller", cause: cause}
}

func sleepContext(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return ctx.Err()
	}

	timer := time.NewTimer(d)
	defer timer.Stop()

	select {
	case <-timer.C:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
